contracts/RemmeBridge.sol

```diff
@@ -1,10 +1,240 @@
+    }
+
+    //FUNCTIONS
+    /// @notice Get swap main info (addresses, state)
+    function getSwapInfo(bytes32 _swapId) view returns (
+        address sender,
+        address receiver,
+        address keyHolder,
+        bytes remChainAddress,
+        State state)
+    {
+        AtomicSwap storage swap = swaps[_swapId];
+        sender = swap.senderAddress;
+        receiver = swap.receiverAddress;
+        keyHolder = swap.keyHolderAddress;
+        remChainAddress = swap.remchainAddress;
+        state = swapStates[_swapId];
+    }
+
+    /// @notice Get swap details info
+    function getSwapDetails(bytes32 _swapId) view returns (
+        uint amount,
+        bytes emailEncrypted,
+        bytes32 secretLock,
+        bytes32 secretKey,
+        uint timelock)
+    {
+	    AtomicSwap storage swap = swaps[_swapId];
+        amount = swap.amount;
+        emailEncrypted = swap.emailAddressEncryptedOptional;
+        secretLock = swap.secretLock;
+        secretKey = swap.secretKey;
+        timelock = swap.timelock;
+    }
+
+    /*
+	* @notice User should use this function for request swap locking his tokens
+	* @dev There are two cases of using this function:
+	* 1) Alice requests swap and don't put _secretLock argument, and sends ether
+	*    required by atomicSwapProvider for gas coverage
+	* 2) Bob opens swap after Alice's request in Remmechain and set _secretLock
+	* @param _secretLock (!) will be set only by Bob. Should be checked in client side during validation
+    */
+    function openSwap (
+        bytes32 _swapId,
+        address _receiverAddress,
+        bytes32 _secretLock,
+        uint256 _amount,
+        bytes _remchainAddress,
+        bytes _emailAddressEncryptedOptional)
+    external
+    {
+	    require(swaps[_swapId].swapId == 0);
+
+        //Bob always set as keyHolder
+	    uint256 lock;
+	    address keyHolder;
+	    uint256 amountToSwap = _amount;
+        if (_secretLock == bytes32(0)) {
+            lock = now + LOCK24;
+            keyHolder = _receiverAddress;
+            if (_receiverAddress == atomicSwapProvider) {
+                require(REMToken.transferFrom(msg.sender, coldStorage, providerFee));
+	            amountToSwap = _amount.sub(providerFee);
+            }
+        } else {
+            lock = now + LOCK48;
+            keyHolder = msg.sender;
+        }
+
+	    AtomicSwap memory swap = AtomicSwap({
+            swapId: _swapId,
+            senderAddress: msg.sender,
+            receiverAddress: _receiverAddress,
+            keyHolderAddress: keyHolder,
+            remchainAddress: _remchainAddress,
+            amount: amountToSwap,
+            emailAddressEncryptedOptional: _emailAddressEncryptedOptional,
+            secretLock: _secretLock,
+            secretKey: "", //will be set only on closing swap
+            timelock: lock
+            });
+        swaps[swap.swapId] = swap;
+
+	    require(REMToken.transferFrom(msg.sender, address(this), amountToSwap));
```

Comment:
Never ever store money on the smart-contract account. A designated external wallet is just fine.

Reply:
There is storing only locked funds during swap, so contract needs to have control

Reply:
Participant should receive fee only if it is bot (provider). There is no additional field for service fee amount, and we need to assume, that it not returnable. Thus they transfer to cold storage instantly. But at this moment swap not closed yet, so funds should be locked and stored on contract address for close or expire request. After closing swap by receiver, the funds will be transferred respectively to address or cold storage

Reply:
So the suggestion is to lock swapping funds on the cold address (instead of the contract given it is a potential point of controls failure - controlling funds by a smart-contract or an agent) and by the time it is closed or expired, transfer back the amount without fees.

Reply:
The cold storage is controlled by third party (provider owners), and it is just e.g. multisig wallet. And swap contract address controlled only by code, and no one can get locked funds if there will be no bugs or logical mistakes. If we give control of cold storage to swap contract, then there will be no difference with that we store funds on swap contract